url901cc5721b_shidasan_KonohaCoffee_tgz-pJ8-org_KonohaScript_KonohaJ8/src/org/KonohaScript/GrammarSet/MiniKonoha.java

```diff
@@ -0,0 +1,1614 @@
+		ParsedTokenList.add(Token);
+		return pos + 1;
+	}
+
+	public int SymbolToken(KNameSpace ns, String SourceText, int pos, ArrayList<KToken> ParsedTokenList) {
+		int start = pos;
+		for (; pos < SourceText.length(); pos++) {
+			char ch = SourceText.charAt(pos);
+			if(!Character.isLetter(ch) && !Character.isDigit(ch) && ch != '_') {
+				break;
+			}
+		}
+		KToken Token = new KToken(SourceText.substring(start, pos));
+		ParsedTokenList.add(Token);
+		return pos;
+	}
+
+	public int MemberToken(KNameSpace ns, String SourceText, int pos, ArrayList<KToken> ParsedTokenList) {
+		int start = pos + 1;
+		for (; pos < SourceText.length(); pos++) {
+			char ch = SourceText.charAt(pos);
+			if(!Character.isLetter(ch) && !Character.isDigit(ch) && ch != '_') {
+				break;
+			}
+		}
+		KToken Token = new KToken(SourceText.substring(start, pos));
+		Token.ResolvedSyntax = KSyntax.MemberSyntax;
+		ParsedTokenList.add(Token);
+		return pos;
+	}
+
+	public int NumberLiteralToken(KNameSpace ns, String SourceText, int pos, ArrayList<KToken> ParsedTokenList) {
+		int start = pos;
+		for (; pos < SourceText.length(); pos++) {
+			char ch = SourceText.charAt(pos);
+			if(!Character.isDigit(ch)) {
+				break;
+			}
+		}
+		KToken token = new KToken(SourceText.substring(start, pos));
+		token.ResolvedSyntax = ns.GetSyntax("$IntegerLiteral");
+		ParsedTokenList.add(token);
+		return pos;
+	}
+
+	public int StringLiteralToken(KNameSpace ns, String SourceText, int pos, ArrayList<KToken> ParsedTokenList) {
+		int start = pos + 1;
+		char prev = '"';
+		pos = start;
+		while (pos < SourceText.length()) {
+			char ch = SourceText.charAt(pos);
+			if (ch == '"' && prev == '\\') {
+				KToken token = new KToken(SourceText.substring(start, pos - start));
+				token.ResolvedSyntax = ns.GetSyntax("$StringLiteral");
+				ParsedTokenList.add(token);
+				return pos + 1;
+			}
+			if (ch == '\n') {
+				KToken token = new KToken(SourceText.substring(start, pos - start));
+				ns.Message(Error, token, "expected \" to close the string literal");
+				ParsedTokenList.add(token);
+				return pos;
+			}
+			pos = pos + 1;
+			prev = ch;
+		}
+		return pos;
+	}
+
+	// Macro
+
+	public int OpenParenthesisMacro(LexicalConverter Lexer, ArrayList<KToken> SourceList, int BeginIdx, int EndIdx, ArrayList<KToken> BufferList) {
+		ArrayList<KToken> GroupList = new ArrayList<KToken>();
+		KToken BeginToken = SourceList.get(BeginIdx);
+		GroupList.add(BeginToken);
+		int nextIdx = Lexer.Do(SourceList, BeginIdx + 1, EndIdx, GroupList);
+		KToken LastToken = GroupList.get(GroupList.size()-1);
+		if (!LastToken.EqualsText(")")) { // ERROR
+			LastToken.SetErrorMessage("must close )");
+		}
+		else {
+			KToken GroupToken = new KToken("( ... )", BeginToken.uline);
+			GroupToken.SetGroup(Lexer.GetSyntax("()"), GroupList);
+			BufferList.add(GroupToken);
+		}
+		//KonohaDebug.P("BeginIdx=" + BeginIdx + ",nextIdx="+nextIdx + ",EndIdx="+EndIdx);
+		return nextIdx;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 134:148, 161:174, 211:224.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.